File: Sorting/quick.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>
#include<math.h>

#define MAX 200000

void swap(int *a, int *b){
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int arr[], int l, int r){
    int x=l, y=r;
    int temp;
    int pivot = arr[l];
    while(x < y){
        while(arr[x] <= pivot && x<=r){
            ++x;
        }
        while(arr[y] > pivot && y >=l){
            --y;
        }
        if(x < y){
            temp = arr[x];
            arr[x] = arr[y];
            arr[y] = temp;
        }
    }

   
    arr[l] = arr[y];
    arr[y] = pivot;
    return y;
}

void quick_sort(int arr[], int l, int r){
    if(l<r){
        int p = partition(arr, l, r);
        quick_sort(arr, l, p-1);
        quick_sort(arr, p+1, r);
    }

}


void display(int arr[], int n){
    int i;
    for(i=0; i<n; i++){
        printf("%d\t", arr[i]);
    }
    printf("\n");
}

int main(){
    int i, a[MAX], n;

    srand(time(NULL));
    time_t t1, t2;
    double d;

    printf("Enter n: ");
    scanf("%d", &n);
    for(i=0; i<n; i++){
        a[i] = rand()%1000;
    }
    display(a, n);
    t1 = time(NULL);
    quick_sort(a, 0, n-1);
    t2 = time(NULL);
    display(a, n);

    d = difftime(t2, t1);

    printf("Total time taken: %lf seconds.\n", d);

    return 0;
}

File: Sorting/quick.py
def partition(a, l, r):
    x, y = l, r
    pivot = a[l]

    while x<y:
        while x<=r and a[x] <= pivot:
            x+=1
        while a[y] > pivot:
            y-=1
        
        if(x < y):
            a[x], a[y] = a[y], a[x]

    a[l], a[y] = a[y], a[l]  
    return y

def quick_sort(a, l, r):
    if l<r:
        p = partition(a, l, r)
        quick_sort(a, l, p-1)
        quick_sort(a, p+1, r)
        

a = [5, 27, 10, 3, 0, 1, -15, 99, 8, 2]
print(a)
quick_sort(a, 0, 9)
print(a)